### DISCUSSION TITLE
Attempted to add ZBL training in finetune but failed

### DISCUSSION THREAD

**Discussion by JiangXiaoMingSan** [deepmodeling/deepmd-kit#4332] [Q&A]
When using deepmdkit-b3 and scripts from GitHub for finetune, an error is reported. When using the se_2_a descriptor, an error is reported: ValueError: LinearAtomicModel only supports AtomicModel of mixed type, the following models are not mixed type: [DPAtomicModel(
  (descriptor): DescrptSeA(
    (sea): DescrptBlockSeA(
      (emask): PairExcludeMask()
      (filter_layers): NetworkCollection(
        (networks): ModuleList(
          (0-1): 2 x EN(
            (layers): ModuleList(
              (0-2): 3 x MLPLayer(
                (activate): ActivationFn()
              )
            )
          )
        )
        (_networks): ModuleList(
  
        (0-1): 2 x EN(
            (layers): ModuleList(
              (0-2): 3 x MLPLayer(
                (activate): ActivationFn()
              )
            )
          )
        )
      )
    )
  )
  (fitting_net): EnergyFittingNet(
    (emask): AtomExcludeMask()
    (filter_layers): NetworkCollection(
      (networks): ModuleList(
        (0-1): 2 x FN(
          (layers): ModuleList(
            (0-3): 4 x MLPLayer(
              (activate): ActivationFn()
            )
          )
        )
      )
      (_networks): ModuleList(
        (0-1): 2 x FN(
          (layers): ModuleList(
            (0-3): 4 x MLPLayer(
              (activate): ActivationFn()
            )
          )
        )
      )
    )
  )
)].

I don't quite understand this error, the model I am using is OpenLAM_2.2.0_27heads_beta3.pt, What type is a mixed type? 
But when I tried to change the descriptor to hybrid and dpa2, I encountered another error about rcut: 
Traceback (most recent call last):
  File "/opt/deepmd-kit-3.0.0b3/bin/dp", line 10, in <module>
    sys.exit(main())
  File "/opt/deepmd-kit-3.0.0b3/lib/python3.10/site-packages/deepmd/main.py", line 923, in main
    deepmd_main(args)
  File "/opt/deepmd-kit-3.0.0b3/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 346, in wrapper
    return f(*args, **kwargs)
  File "/opt/deepmd-kit-3.0.0b3/lib/python3.10/site-packages/deepmd/pt/entrypoints/main.py", line 562, in main
    train(FLAGS)
  File "/opt/deepmd-kit-3.0.0b3/lib/python3.10/site-packages/deepmd/pt/entrypoints/main.py", line 326, in train
    trainer = get_trainer(
  File "/opt/deepmd-kit-3.0.0b3/lib/python3.10/site-packages/deepmd/pt/entrypoints/main.py", line 190, in get_trainer
    trainer = training.Trainer(
  File "/opt/deepmd-kit-3.0.0b3/lib/python3.10/site-packages/deepmd/pt/train/training.py", line 277, in __init__
    self.model = get_model_for_wrapper(model_params)
  File "/opt/deepmd-kit-3.0.0b3/lib/python3.10/site-packages/deepmd/pt/train/training.py", line 1258, in get_model_for_wrapper
    _model = get_single_model(
  File "/opt/deepmd-kit-3.0.0b3/lib/python3.10/site-packages/deepmd/pt/train/training.py", line 1250, in get_single_model
    model = get_zbl_model(deepcopy(_model_params)).to(DEVICE)
  File "/opt/deepmd-kit-3.0.0b3/lib/python3.10/site-packages/deepmd/pt/model/model/__init__.py", line 131, in get_zbl_model
    model_params["descriptor"]["rcut"],
KeyError: 'rcut'

 Specific scripts can be found on GitHub, and I have added ZBL to all the scripts.( [https://github.com/JiangXiaoMingSan/finetune_zbl_test.git](https://github.com/JiangXiaoMingSan/finetune_zbl_test.git) )

**Comment by njzjz** [comment]
> File "/opt/deepmd-kit-3.0.0b3/lib/python3.10/site-packages/deepmd/pt/model/model/**init**.py", line 131, in get_zbl_model
> model_params["descriptor"]["rcut"],
> KeyError: 'rcut'

@anyangml please take a look at this issue. The descriptor should not be assumed to be dpa1

**Reply by njzjz** [reply]
`descriptor.get_rcut()` and `descriptor.get_sel()` should be used instead.

**Reply by anyangml** [reply]
In the example we provided, we did not give the pairtable model a separate `rcut` and `sel` . We simply used the the `rcut` and `sel` from the dp model. Technically, [pairtable model](https://github.com/deepmodeling/deepmd-kit/blob/dcbf607f5544d1269d32837f9d045a49234ad910/deepmd/pt/model/atomic_model/pairtab_atomic_model.py#L68) has its own `rcut` and `sel`, and when combining into a ZBL model, the `rcut` and `sel` are processed within the [ZBL model](https://github.com/deepmodeling/deepmd-kit/blob/dcbf607f5544d1269d32837f9d045a49234ad910/deepmd/pt/model/atomic_model/linear_atomic_model.py#L154). If we want the flexibility of using a different `rcut` and `sel`, we would need to add these information in the `input.json` at the `model_params` level. If we don't need this flexibility, we can simply update the get_zbl_model method.

**Comment by anyangml** [comment]
fixed in https://github.com/deepmodeling/deepmd-kit/pull/4335
